### PR TITLE
Add tests for RegExp.escape

### PR DESCRIPTION
**Pull request by leobalter**
Initial implementation of tests for RegExp.escape. I still need to review if I'm missing coverage here or running the tests.

### CODE REVIEW

File: test/built-ins/RegExp/escape/escaped-code-points.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// Copyright (C) 2024 Leo Balter. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-encodeforregexescape
+description: Encodes code points greater than 0xFF
```
Comment (bakkot) [inline test/built-ins/RegExp/escape/escaped-code-points.js:6]: I think this is wrong. None of the conditions in step 5 apply to any of the listed characters, so they fall through to step 6, which just returns the characters without escaping them.

File: test/built-ins/RegExp/escape/escaped-syntax-characters-mixed.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-regexp.escape
+description: Escaped characters (mixed assertions)
+info: |
+  RegExp.escape ( string )
+
+  This method produces a new string in which certain characters have been escaped.
+  These characters are: . * + ? ^ $ | ( ) [ ] { } \
```
Comment (bakkot) [inline test/built-ins/RegExp/escape/escaped-syntax-characters-mixed.js:11]: Also each of

```
, - = < > # & ! % : ; @ ~ ' ` "
```
per step 3, and `/` per step 1 (the `c is U+002F (SOLIDUS)` case).

Test below asserts that those are not escaped, but that's wrong. They're a different branch for editorial reasons, but they're still escaped just the same as these.

**Comment by leobalter** [comment]
Thanks @bakkot! I fixed some coverage. I'll have a hard stop for now but will continue later.

**Comment by leobalter** [comment]
This should be ready for another round of reviews.

File: test/built-ins/RegExp/escape/initial-char-escape.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+assert.sameValue(RegExp.escape('3+3'), '\\x33\\+3', 'Initial decimal digit 3 with special character is escaped');
+assert.sameValue(RegExp.escape('4+4'), '\\x34\\+4', 'Initial decimal digit 4 with special character is escaped');
+assert.sameValue(RegExp.escape('5+5'), '\\x35\\+5', 'Initial decimal digit 5 with special character is escaped');
+assert.sameValue(RegExp.escape('6+6'), '\\x36\\+6', 'Initial decimal digit 6 with special character is escaped');
+assert.sameValue(RegExp.escape('7+7'), '\\x37\\+7', 'Initial decimal digit 7 with special character is escaped');
+assert.sameValue(RegExp.escape('8+8'), '\\x38\\+8', 'Initial decimal digit 8 with special character is escaped');
+assert.sameValue(RegExp.escape('9+9'), '\\x39\\+9', 'Initial decimal digit 9 with special character is escaped');
+assert.sameValue(RegExp.escape('0+0'), '\\x30\\+0', 'Initial decimal digit 0 with special character is escaped');
+
+assert.sameValue(RegExp.escape('a*a'), '\\x62\\*a', 'Initial ASCII letter a with special character is escaped');
```
Comment (ljharb) [inline test/built-ins/RegExp/escape/initial-char-escape.js:105]: why would the presence of the `*` cause the letter before it to be escaped?

File: test/built-ins/RegExp/escape/not-escaped-underscore.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  This method produces a new string in which certain characters have been escaped.
+  These characters are: . * + ? ^ $ | ( ) [ ] { } \ , - = < > # & ! % : ; @ ~ ' ` " and white space or line terminators.
+features: [RegExp.escape]
+---*/
+
+assert.sameValue(RegExp.escape('_'), '_', 'Single underscore character is not escaped');
+assert.sameValue(RegExp.escape('__'), '__', 'Thunderscore character is not escaped');
+assert.sameValue(RegExp.escape('hello_world'), '\\x68ello_world', 'String starting with ASCII letter and containing underscore is correctly escaped');
+assert.sameValue(RegExp.escape('1_hello_world'), '\\x31_hello_world', 'String starting with digit and containing underscore is correctly escaped');
+assert.sameValue(RegExp.escape('a_b_c'), '\\x61_b_c', 'String starting with ASCII letter and containing multiple underscores is correctly escaped');
```
Comment (ljharb) [inline test/built-ins/RegExp/escape/not-escaped-underscore.js:19]: the first character is only escaped differently if it's a decimal digit
```suggestion
assert.sameValue(RegExp.escape('a_b_c'), 'a_b_c', 'String starting with ASCII letter and containing multiple underscores is not escaped');
```

File: test/built-ins/RegExp/escape/not-escaped-underscore.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  RegExp.escape ( string )
+
+  This method produces a new string in which certain characters have been escaped.
+  These characters are: . * + ? ^ $ | ( ) [ ] { } \ , - = < > # & ! % : ; @ ~ ' ` " and white space or line terminators.
+features: [RegExp.escape]
+---*/
+
+assert.sameValue(RegExp.escape('_'), '_', 'Single underscore character is not escaped');
+assert.sameValue(RegExp.escape('__'), '__', 'Thunderscore character is not escaped');
+assert.sameValue(RegExp.escape('hello_world'), '\\x68ello_world', 'String starting with ASCII letter and containing underscore is correctly escaped');
```
Comment (ljharb) [inline test/built-ins/RegExp/escape/not-escaped-underscore.js:17]: the first character is only escaped differently if it's a decimal digit
```suggestion
assert.sameValue(RegExp.escape('hello_world'), 'hello_world', 'String starting with ASCII letter and containing underscore is not escaped');
```

**Review by ljharb — COMMENTED** [review]
I haven't tried the two files that @bakkot commented on yet

File: test/built-ins/RegExp/escape/not-escaped-underscore.js (hunk 5 above)
Comment (zloirock) [reply test/built-ins/RegExp/escape/not-escaped-underscore.js:17]: > `1. If _escaped_ is the empty String, and _c_ is matched by |DecimalDigit| or |AsciiLetter|, then`

Comment (ljharb) [reply test/built-ins/RegExp/escape/not-escaped-underscore.js:17]: oh whoops, you're right. thanks, i'll update my polyfill.

File: test/built-ins/RegExp/escape/initial-char-escape.js (hunk 3 above)
Comment (leobalter) [reply test/built-ins/RegExp/escape/initial-char-escape.js:105]: this is just a little convenience to mix the first digit escaping with non-escaped and escaped codepoints. The first group of assertions line 28-91 only use decimaldigits and ascii in the strings. The other following assertions mix up the first escaped code point with something else. I subjectively picked `+` and `*` just for sake of anything else and common characters used in regex.

File: test/built-ins/RegExp/escape/escaped-whitespace.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+---*/
+
+const WhiteSpace = '\uFEFF\u0020\u00A0\u202F';
+
+assert.sameValue(RegExp.escape('\uFEFF'), '\\uFEFF', `whitespace \\uFEFF is escaped correctly to \\uFEFF`);
+assert.sameValue(RegExp.escape('\u0020'), '\\x20', `whitespace \\u0020 is escaped correctly to \\x20`);
+assert.sameValue(RegExp.escape('\u00A0'), '\\xA0', `whitespace \\u00A0 is escaped correctly to \\xA0`);
+assert.sameValue(RegExp.escape('\u202F'), '\\u202F', `whitespace \\u202F is escaped correctly to \\u202F`);
+
+assert.sameValue(RegExp.escape(WhiteSpace), '\\uFEFF\\x20\\xA0\\u202F', `whitespaces are escaped correctly`);
```
[the remark this answers is not in the document]
Comment (leobalter) [reply test/built-ins/RegExp/escape/escaped-whitespace.js:55]: interesting, it took me some time to find yesterday but indeed some are codes over `0xFF` are lower cased as they are parsed through [UnicodeEscape](https://tc39.es/ecma262/#sec-unicodeescape).

This is valid for this test and surrogates as anything falling through step 5.d.i on EncodeForRegExpEscape.

the same does not apply to others through step 6 and codes are not lower cased, correct?

File: test/built-ins/RegExp/escape/escaped-surrogates.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+---*/
+
+// Specific surrogate points
+assert.sameValue(RegExp.escape('\uD800'), '\\uD800', 'High surrogate \\uD800 is correctly escaped');
+assert.sameValue(RegExp.escape('\uDBFF'), '\\uDBFF', 'High surrogate \\uDBFF is correctly escaped');
+assert.sameValue(RegExp.escape('\uDC00'), '\\uDC00', 'Low surrogate \\uDC00 is correctly escaped');
+assert.sameValue(RegExp.escape('\uDFFF'), '\\uDFFF', 'Low surrogate \\uDFFF is correctly escaped');
+
+// Leading Surrogates
+const highSurrogatesGroup1 = '\uD800\uD801\uD802\uD803\uD804\uD805\uD806\uD807\uD808\uD809\uD80A\uD80B\uD80C\uD80D\uD80E\uD80F';
```
[the remark this answers is not in the document]
Comment (leobalter) [reply test/built-ins/RegExp/escape/escaped-surrogates.js:30]: nit pick: not due to number-tostring but [unicodeescape](https://tc39.es/ecma262/#sec-unicodeescape)

File: test/built-ins/RegExp/escape/escaped-utf16encodecodepoint..js [hunk 8]
```diff
@@ -0,0 +1,40 @@
+// Copyright (C) 2024 Leo Balter. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
```
[the remark this answers is not in the document]
Comment (leobalter) [reply test/built-ins/RegExp/escape/escaped-utf16encodecodepoint..js]: this commit fixed a surrogate issue 14e646a. Is it working now?

File: test/built-ins/RegExp/escape/initial-char-escape.js [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+      v. Set escaped to the string-concatenation of the code unit 0x005C (REVERSE SOLIDUS), "x", and hex.
+    b. Else,
+      i. Set escaped to the string-concatenation of escaped and EncodeForRegExpEscape(c).
+  5. Return escaped.
+
+features: [RegExp.escape]
+---*/
+
+// Escaping initial digits
+assert.sameValue(RegExp.escape('1111'), '\\x31111', 'Initial decimal digit 1 is escaped');
```
[the remark this answers is not in the document]
Comment (leobalter) [reply test/built-ins/RegExp/escape/initial-char-escape.js:28]: I believe this got fixed during the review. PTAL

File: test/built-ins/RegExp/escape/escaped-surrogates.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (leobalter) [reply test/built-ins/RegExp/escape/escaped-surrogates.js]: fixed in [e00d756](https://github.com/tc39/test262/pull/4123/commits/e00d7567ff75948ab9ebaaf6a3c4468e25532615)

File: test/built-ins/RegExp/escape/escaped-solidus-character-mixed.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  EncodeForRegExpEscape ( c )
+
+  1. If c is matched by SyntaxCharacter or c is U+002F (SOLIDUS), then
+    a. Return the string-concatenation of 0x005C (REVERSE SOLIDUS) and UTF16EncodeCodePoint(c).
+features: [RegExp.escape]
+---*/
+
+assert.sameValue(RegExp.escape('.a/b'), '\\.a\\/b', 'mixed string with solidus character is escaped correctly');
+assert.sameValue(RegExp.escape('/./'), '\\/\\.\\/', 'solidus character is escaped correctly - regexp similar');
+assert.sameValue(RegExp.escape('./a\\/*b+c?d^e$f|g{2}h[i]j\\k'), '\\.\\/a\\\\/\\*b\\+c\\?d\\^e\\$f\\|g\\{2\\}h\\[i\\]j\\\\k', 'complex string with multiple special characters is escaped correctly');
```
[the remark this answers is not in the document]
Comment (leobalter) [reply test/built-ins/RegExp/escape/escaped-solidus-character-mixed.js]: I believe I was missing a `\` escaping there. I fixed it in [2efa69c](https://github.com/tc39/test262/pull/4123/commits/2efa69c55fc5dccd1e77dfa3d111f72d2901caff), PTAL

**Comment by leobalter** [comment]
Thanks, @ljharb! I believe I addressed all the review points so far.

File: test/built-ins/RegExp/escape/escaped-code-points.js (hunk 1 above)
Comment (ljharb) [reply test/built-ins/RegExp/escape/escaped-code-points.js:6]: agree, this test seems wrong to me.

**Review by ljharb — APPROVED** [review]
Took over the PR at Leo's request; PTAL

**Comment by ljharb** [comment]
Are these ready to merge? I'm hoping to ask for stage 3 at this month's plenary.

**Review by Ms2ger — APPROVED** [review]

**Pull request merged**